src/libraries/System.Net.Http/tests/FunctionalTests/Socks/LoopbackSocksServer.cs

```diff
@@ -0,0 +1,365 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.IO;
+using System.Net.Sockets;
+using System.Net.Test.Common;
+using System.Runtime.ExceptionServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Functional.Tests.Socks
+{
+    /// <summary>
+    /// Provides a test-only SOCKS4/5 proxy.
+    /// </summary>
+    internal class LoopbackSocksServer : IDisposable
+    {
+        private readonly Socket _listener;
+        private readonly ManualResetEvent _serverStopped;
+        private bool _disposed;
+
+        private int _connections;
+        public int Connections => _connections;
+
+        public int Port { get; }
+
+        private string? _username, _password;
+
+        private LoopbackSocksServer(string? username = null, string? password = null)
+        {
+            if (password != null && username == null)
+            {
+                throw new ArgumentException("Password must be used together with username.", nameof(password));
```

Comment:
Is it valid to have a non-null username and a null password?

Reply:
Yes in case of socks4(a), which only uses the username as an identifier, but doesn't actually do user/pass auth.

Reply:
rfc1929 spec implies username and password shouldn't be 0 length. Should we add a guard for that? Looks overkill.

Reply:
I wouldn't bother, we can let the server decide whether to reject it.